# Regression Cross-Validation

In this notebook, we apply the cross-validation workflow to regression.

We will use the California Housing dataset and evaluate Linear Regression using:

- R²
- MAE
- RMSE

In [21]:
from sklearn.datasets import fetch_california_housing

data = fetch_california_housing(as_frame=True)

df = data.frame

X = df.drop(columns="MedHouseVal")
y = df["MedHouseVal"]

df.head()

,MedInc,HouseAge,AveRooms,AveBedrms,Population,AveOccup,Latitude,Longitude,MedHouseVal
0,8.3252,41.0,6.984127,1.023810,322.0,2.555556,37.88,-122.23,4.526
1,8.3014,21.0,6.238137,0.971880,2401.0,2.109842,37.86,-122.22,3.585
2,7.2574,52.0,8.288136,1.073446,496.0,2.802260,37.85,-122.24,3.521
3,5.6431,52.0,5.817352,1.073059,558.0,2.547945,37.85,-122.25,3.413
4,3.8462,52.0,6.281853,1.081081,565.0,2.181467,37.85,-122.25,3.422


In [22]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(
    X,
    y,
    test_size=0.2,
    random_state=42,
)

In [23]:
from sklearn.linear_model import LinearRegression

model = LinearRegression()

# R² Cross-Validation

R² is already a "higher is better" score, so we can use `scoring="r2"` directly.

In [24]:
from sklearn.model_selection import cross_val_score

r2_scores = cross_val_score(
    model,
    X_train,
    y_train,
    cv=5,
    scoring="r2",
)

r2_scores

array([0.62011512, 0.61298876, 0.6134416 , 0.61069973, 0.60017477])

In [25]:
print(f"Mean CV R²: {r2_scores.mean():.4f}")
print(f"Std CV R²: {r2_scores.std():.4f}")

Mean CV R²: 0.6115
Std CV R²: 0.0065


# MAE Cross-Validation

MAE is an error metric where lower is better.

Because scikit-learn's scoring convention expects higher scores to be better, we use:

`neg_mean_absolute_error`

The returned values are negative.

We convert them back to positive MAE values for interpretation.

In [26]:
mae_scores = cross_val_score(
    model,
    X_train,
    y_train,
    cv=5,
    scoring="neg_mean_absolute_error",
)

mae_scores

array([-0.53113061, -0.52675118, -0.52419835, -0.5247937 , -0.5384344 ])

In [27]:
print(f"Mean CV MAE: {-mae_scores.mean():.4f}")
print(f"Std CV MAE: {mae_scores.std():.4f}")

Mean CV MAE: 0.5291
Std CV MAE: 0.0053


# RMSE Cross-Validation

RMSE is also an error metric where lower is better.

Therefore, scikit-learn provides the scoring name:

`neg_root_mean_squared_error`

In [28]:
rmse_scores = cross_val_score(
    model,
    X_train,
    y_train,
    cv=5,
    scoring="neg_root_mean_squared_error",
)

print(f"Mean CV RMSE: {-rmse_scores.mean():.4f}")
print(f"Std CV RMSE: {rmse_scores.std():.4f}")

Mean CV RMSE: 0.7205
Std CV RMSE: 0.0103


In [29]:
import pandas as pd

results = pd.DataFrame({
    "Metric": ["R²", "MAE", "RMSE"],
    "Mean CV Score": [
        r2_scores.mean(),
        -mae_scores.mean(),
        -rmse_scores.mean(),
    ],
    "Std": [
        r2_scores.std(),
        mae_scores.std(),
        rmse_scores.std(),
    ],
})

results

,Metric,Mean CV Score,Std
0,R²,0.611484,0.006467
1,MAE,0.529062,0.005279
2,RMSE,0.720527,0.010285


# Interpreting the Results

For R²:

Higher is better.

For MAE and RMSE:

Lower is better.

The standard deviation shows how much the score varied across the five folds.

The final test set was not used during this cross-validation process.

# Interpreting the Cross-Validation Results

The Linear Regression model produced:

- Mean CV R² = 0.6115
- Mean CV MAE = 0.5291
- Mean CV RMSE = 0.7205

The R² score indicates that the model explains approximately 61.1% of the target variation across the validation folds.

The mean MAE is approximately 0.529 target units.

The mean RMSE is approximately 0.721 target units.

The relatively small standard deviations indicate that the model's performance was fairly consistent across the five folds.

These CV results should not be directly compared with the single test-set metrics from the previous notebook because they come from different evaluation procedures.

# Key Takeaways

- Cross-validation can be used for regression as well as classification.
- R² can be used directly as a scoring metric.
- MAE and RMSE use negative scoring names in scikit-learn.
- Negative MAE/RMSE scores must be converted back to positive error values for interpretation.
- Mean CV score summarizes performance across folds.
- Standard deviation shows variation across folds.
- A reserved test set should remain untouched during model selection.

# Notebook Summary

In this notebook, we applied cross-validation to a regression problem.

We evaluated Linear Regression using:

- R²
- MAE
- RMSE

We also learned how scikit-learn represents regression losses as negative scores during cross-validation.

# What's Next?

Next: Pipelines and Data Leakage.

This is one of the most important practical topics before starting the project.

We will learn how preprocessing can accidentally leak information from validation or test data into training, and how Pipelines help prevent this.